File: Python/Graphs/BronKerbosch.py
#   The Bron-Kerbosch is a algorithm to find all maximal cliques (complete subgraphs) in an undirected graph.
#   Clique: Set of vertices such that every pair of vertices is connected by an edge.
#   Maximal Clique: A clique is maximal if it cannot be extended by adding another vertex in the graph.

def bron_kerbosch(clique, potential, excluded, graph):
    if not potential and not excluded:
        print("Maximal Clique:", end=" ")
        print(*clique)
        return

    potential_copy = potential.copy()
    for vertex in potential_copy:
        new_clique = clique.copy()
        new_clique.add(vertex)
        new_potential = set()
        new_excluded = set()

        for neighbor in potential:
            if graph[vertex][neighbor] == 1:
                new_potential.add(neighbor)
        for neighbor in excluded:
            if graph[vertex][neighbor] == 1:
                new_excluded.add(neighbor)

        bron_kerbosch(new_clique, new_potential, new_excluded, graph)

        potential.remove(vertex)
        excluded.add(vertex)


if __name__ == "__main__":
    N = int(input("Enter the number of vertices: "))

    graph = [[0] * N for _ in range(N)]
    print("Enter the adjacency matrix:")
    for i in range(N):
        graph[i] = list(map(int, input().split()))

    clique = set()
    potential = set(range(N))
    excluded = set()

    bron_kerbosch(clique, potential, excluded, graph)

File: Java/Graphs/BronKerbosch.java
/*
 *   The Bron-Kerbosch is a algorithm to find all maximal cliques (complete subgraphs) in an undirected graph.
 *   Clique: Set of vertices such that every pair of vertices is connected by an edge.
 *   Maximal Clique: A clique is maximal if it cannot be extended by adding another vertex in the graph.
 */
import java.util.*;

public class BronKerbosch {

    public static void bronKerbosch(Set<Integer> clique, Set<Integer> potential, Set<Integer> excluded, int[][] graph) {
        if (potential.isEmpty() && excluded.isEmpty()) {
            System.out.print("Maximal Clique: ");
            for (int vertex : clique) {
                System.out.print(vertex + " ");
            }
            System.out.println();
            return;
        }

        Set<Integer> potentialCopy = new HashSet<>(potential);
        for (int vertex : potentialCopy) {
            Set<Integer> newClique = new HashSet<>(clique);
            newClique.add(vertex);
            Set<Integer> newPotential = new HashSet<>();
            Set<Integer> newExcluded = new HashSet<>();

            for (int neighbor : potential) {
                if (graph[vertex][neighbor] == 1) {
                    newPotential.add(neighbor);
                }
            }
            for (int neighbor : excluded) {
                if (graph[vertex][neighbor] == 1) {
                    newExcluded.add(neighbor);
                }
            }

            bronKerbosch(newClique, newPotential, newExcluded, graph);

            potential.remove(vertex);
            excluded.add(vertex);
        }
    }

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter the number of vertices: ");
            int N = scanner.nextInt();

            int[][] graph = new int[N][N];
            System.out.println("Enter the adjacency matrix:");
            for (int i = 0; i < N; ++i) {
                for (int j = 0; j < N; ++j) {
                    graph[i][j] = scanner.nextInt();
                }
            }

            Set<Integer> clique = new HashSet<>();
            Set<Integer> potential = new HashSet<>();
            Set<Integer> excluded = new HashSet<>();
            for (int i = 0; i < N; ++i) {
                potential.add(i);
            }

            bronKerbosch(clique, potential, excluded, graph);
        }
    }
}